File: python/src/bench_sdk/platform.py
"""Headless Bench management. Explicit operations; no implicit retries or spending."""
from __future__ import annotations
import json
from importlib.resources import files
from urllib.parse import urlsplit, quote, urlencode
from urllib.request import Request, build_opener, HTTPRedirectHandler
from urllib.error import HTTPError
from uuid import uuid4

class PlatformError(Exception):
    def __init__(self, status, code, message, reference=None, details=None):
        super().__init__(message)
        self.status, self.code, self.reference = status, code, reference
        self.details = details or {}
    @property
    def retryable(self):
        return self.status == 429 or self.status >= 500

class _NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None

class BenchPlatform:
    """Use a key for scoped data work or a Bench OAuth credential for account setup.

    credential can be a callable supplying refreshed tokens. Calls are synchronous;
    run local application evaluation with Bench separately, then publish its report.
    """
    def __init__(self, credential=None, *, endpoint="https://api.usebench.ai", timeout=120):
        url = urlsplit(endpoint)
        if url.username or url.password or url.query or url.fragment or url.path not in ("", "/") or not url.hostname or not (url.scheme == "https" or (url.scheme == "http" and url.hostname in ("localhost", "127.0.0.1", "::1"))):
            raise ValueError("Use an HTTPS API origin or loopback HTTP origin.")
        if not isinstance(timeout, (int, float)) or not 0.1 <= timeout <= 600:
            raise ValueError("timeout must be between 0.1 and 600 seconds")
        self._endpoint, self._credential, self._timeout = endpoint.rstrip("/"), credential, timeout
        self._catalog = json.loads(files("bench_sdk").joinpath("operations.json").read_text())
        self._operations = {op["id"]: op for op in self._catalog["operations"]}
        self._opener = build_opener(_NoRedirect())

    def operations(self):
        return json.loads(json.dumps(self._catalog))

    def call(self, operation, *, path=None, query=None, body=None, form=None, files=None):
        """files is a list of {name, content}; content accepts text or bytes."""
        if operation not in self._operations:
            raise ValueError("Unknown Bench operation")
        op = self._operations[operation]
        route = op["path"]
        for key in op["path_parameters"]:
            value = (path or {}).get(key)
            if value is None or not str(value) or str(value) in (".", "..") or any(c.isspace() or c in "/?#\\%" for c in str(value)):
                raise ValueError("Invalid or missing path parameter: " + key)
            route = route.replace("{" + key + "}", quote(str(value), safe=""))
        url = self._endpoint + route
        if query:
            url += "?" + urlencode({k: str(v).lower() if isinstance(v, bool) else v for k, v in query.items() if v is not None})
        credential = None if op["auth"] == "public" else (self._credential() if callable(self._credential) else self._credential)
        headers = {"Accept": "application/json, application/x-ndjson"}
        if op["auth"] != "public":
            if not isinstance(credential, str) or not credential.strip():
                raise PlatformError(401, "unauthorized", "Provide a Bench credential or authenticate through MCP OAuth.")
            headers["Authorization"] = "Bearer " + credential
        payload = None
        if op.get("multipart"):
            if body is not None:
                raise ValueError("Use form and files for multipart operations.")
            boundary = "bench-" + uuid4().hex
            parts = []
            for key, value in (form or {}).items():
                if any(c in str(key) for c in '\r\n"'):
                    raise ValueError("Invalid form field name")
                if value is None:
                    continue
                text = value if isinstance(value, str) else json.dumps(value, separators=(",", ":"))
                parts.append((f'--{boundary}\r\nContent-Disposition: form-data; name="{key}"\r\n\r\n{text}\r\n').encode())
            total = 0
            for file in files or []:
                name, content = file["name"], file["content"]
                if not name or any(c in name for c in '\r\n/\\"'):
                    raise ValueError("Use a file name without directories or line breaks.")
                data = content.encode() if isinstance(content, str) else bytes(content)
                total += len(data)
                if total > op.get("max_file_bytes", 4 * 1024 * 1024):
                    raise ValueError("Upload exceeds the operation file-size limit.")
                parts.append((f'--{boundary}\r\nContent-Disposition: form-data; name="{op.get("file_field", "file")}"; filename="{name}"\r\nContent-Type: application/octet-stream\r\n\r\n').encode() + data + b"\r\n")
            payload = b"".join(parts) + (f"--{boundary}--\r\n").encode()
            headers["Content-Type"] = "multipart/form-data; boundary=" + boundary
        else:
            if files is not None or form is not None:
                raise ValueError("This operation does not accept multipart uploads.")
            if body is not None:
                payload = json.dumps(body, allow_nan=False).encode()
                headers["Content-Type"] = "application/json"
        request = Request(url, data=payload, headers=headers, method=op["method"])
        try:
            response = self._opener.open(request, timeout=self._timeout)
        except HTTPError as error:
            response = error
        with response:
            raw = response.read(16 * 1024 * 1024 + 1)
            if len(raw) > 16 * 1024 * 1024:
                raise PlatformError(response.status, "response_too_large", "Response exceeds 16 MB; use pagination.")
            if not 200 <= response.status < 300:
                try:
                    error = json.loads(raw).get("error", {})
                except (ValueError, AttributeError):
                    error = {}
                raise PlatformError(response.status, error.get("code", "http_" + str(response.status)), error.get("message", "Bench request failed."), error.get("reference"), error)
            if not raw or response.status == 204:
                return None
            if "ndjson" in response.headers.get("Content-Type", ""):
                events = [json.loads(line) for line in raw.splitlines() if line.strip()]
                for event in events:
                    if event.get("type") == "error":
                        raise PlatformError(200, event.get("code", "stream_error"), str(event.get("message", event.get("error", "Operation failed."))))
                return events
            return json.loads(raw)

File: go/platform.go
package bench

import (
	"bytes"
	"context"
	_ "embed"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"strings"
	"time"
	"unicode"
)

//go:embed operations.json
var platformCatalog []byte

type PlatformOptions struct {
	Credential string
	// Token supplies a fresh Bench OAuth/session token for each protected call.
	Token     func(context.Context) (string, error)
	Endpoint  string
	Timeout   time.Duration
	Transport http.RoundTripper
}
type PlatformFile struct {
	Name    string
	Content []byte
}
type PlatformRequest struct {
	Path  map[string]string
	Query url.Values
	Body  any
	Form  map[string]any
	Files []PlatformFile
}
type PlatformError struct {
	Status                   int
	Code, Message, Reference string
	Details                  map[string]any
}

func (e *PlatformError) Error() string   { return e.Message }
func (e *PlatformError) Retryable() bool { return e.Status == 429 || e.Status >= 500 }

type platformOperation struct {
	ID, Method, Path, Auth string
	PathParameters         map[string]json.RawMessage `json:"path_parameters"`
	Multipart              bool
	FileField              string `json:"file_field"`
	MaxFileBytes           int    `json:"max_file_bytes"`
}
type Platform struct {
	options    PlatformOptions
	client     *http.Client
	operations map[string]platformOperation
}

// NewPlatform manages Bench headlessly, independently of application tracing.
// Calls never automatically retry mutations or start an evaluation implicitly.
func NewPlatform(options PlatformOptions) (*Platform, error) {
	if options.Endpoint == "" {
		options.Endpoint = "https://api.usebench.ai"
	}
	u, err := url.Parse(options.Endpoint)
	if err != nil || u.Hostname() == "" || u.User != nil || u.RawQuery != "" || u.Fragment != "" || (u.Path != "" && u.Path != "/") || !(u.Scheme == "https" || (u.Scheme == "http" && (u.Hostname() == "localhost" || u.Hostname() == "127.0.0.1" || u.Hostname() == "::1"))) {
		return nil, errors.New("use an HTTPS API origin or loopback HTTP origin")
	}
	options.Endpoint = strings.TrimRight(options.Endpoint, "/")
	if options.Timeout == 0 {
		options.Timeout = 120 * time.Second
	}
	if options.Timeout < 100*time.Millisecond || options.Timeout > 600*time.Second {
		return nil, errors.New("timeout must be between 100ms and 600s")
	}
	var catalog struct{ Operations []platformOperation }
	if err = json.Unmarshal(platformCatalog, &catalog); err != nil {
		return nil, err
	}
	p := &Platform{options: options, operations: map[string]platformOperation{}, client: &http.Client{Transport: options.Transport, Timeout: options.Timeout, CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }}}
	for _, op := range catalog.Operations {
		p.operations[op.ID] = op
	}
	return p, nil
}
func (p *Platform) Operations() json.RawMessage {
	return append(json.RawMessage(nil), platformCatalog...)
}

// Call maps an operation name from Operations to one HTTP request. For NDJSON
// endpoints it returns the full event array and surfaces in-stream errors.
func (p *Platform) Call(ctx context.Context, operation string, input PlatformRequest) (json.RawMessage, error) {
	op, ok := p.operations[operation]
	if !ok {
		return nil, errors.New("unknown Bench operation")
	}
	route := op.Path
	for key := range op.PathParameters {
		value := input.Path[key]
		if value == "" || value == "." || value == ".." || strings.ContainsAny(value, "/?#\\%") || strings.IndexFunc(value, unicode.IsSpace) >= 0 {
			return nil, fmt.Errorf("invalid or missing path parameter: %s", key)
		}
		route = strings.ReplaceAll(route, "{"+key+"}", url.PathEscape(value))
	}
	endpoint := p.options.Endpoint + route
	if len(input.Query) > 0 {
		endpoint += "?" + input.Query.Encode()
	}
	var payload bytes.Buffer
	contentType := ""
	if op.Multipart {
		if input.Body != nil {
			return nil, errors.New("use form and files for multipart operations")
		}
		writer := multipart.NewWriter(&payload)
		for key, value := range input.Form {
			var text string
			if s, ok := value.(string); ok {
				text = s
			} else {
				raw, err := json.Marshal(value)
				if err != nil {
					return nil, err
				}
				text = string(raw)
			}
			if err := writer.WriteField(key, text); err != nil {
				return nil, err
			}
		}
		total := 0
		limit := op.MaxFileBytes
		if limit == 0 {
			limit = 4 << 20
		}
		field := op.FileField
		if field == "" {
			field = "file"
		}
		for _, file := range input.Files {
			if file.Name == "" || strings.ContainsAny(file.Name, "\r\n/\\") {
				return nil, errors.New("use a file name without directories or line breaks")
			}
			total += len(file.Content)
			if total > limit {
				return nil, errors.New("upload exceeds operation file-size limit")
			}
			part, err := writer.CreateFormFile(field, file.Name)
			if err != nil {
				return nil, err
			}
			if _, err = part.Write(file.Content); err != nil {
				return nil, err
			}
		}
		if err := writer.Close(); err != nil {
			return nil, err
		}
		contentType = writer.FormDataContentType()
	} else {
		if input.Files != nil || input.Form != nil {
			return nil, errors.New("operation does not accept multipart uploads")
		}
		if input.Body != nil {
			raw, err := json.Marshal(input.Body)
			if err != nil {
				return nil, err
			}
			payload.Write(raw)
			contentType = "application/json"
		}
	}
	req, err := http.NewRequestWithContext(ctx, op.Method, endpoint, &payload)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json, application/x-ndjson")
	if contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}
	if op.Auth != "public" {
		credential := p.options.Credential
		if p.options.Token != nil {
			credential, err = p.options.Token(ctx)
			if err != nil {
				return nil, err
			}
		}
		if strings.TrimSpace(credential) == "" {
			return nil, &PlatformError{Status: 401, Code: "unauthorized", Message: "Provide a Bench credential or authenticate through MCP OAuth."}
		}
		req.Header.Set("Authorization", "Bearer "+credential)
	}
	response, err := p.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()
	raw, err := io.ReadAll(io.LimitReader(response.Body, (16<<20)+1))
	if err != nil {
		return nil, err
	}
	if len(raw) > 16<<20 {
		return nil, &PlatformError{Status: response.StatusCode, Code: "response_too_large", Message: "Response exceeds 16 MB; use pagination."}
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		var body struct {
			Error struct{ Code, Message, Reference string }
		}
		_ = json.Unmarshal(raw, &body)
		var envelope struct{ Error map[string]any }
		_ = json.Unmarshal(raw, &envelope)
		if body.Error.Code == "" {
			body.Error.Code = fmt.Sprintf("http_%d", response.StatusCode)
		}
		if body.Error.Message == "" {
			body.Error.Message = "Bench request failed."
		}
		return nil, &PlatformError{Status: response.StatusCode, Code: body.Error.Code, Message: body.Error.Message, Reference: body.Error.Reference, Details: envelope.Error}
	}
	if response.StatusCode == 204 || len(raw) == 0 {
		return json.RawMessage("null"), nil
	}
	if strings.Contains(response.Header.Get("Content-Type"), "ndjson") {
		events := []json.RawMessage{}
		for _, line := range bytes.Split(raw, []byte{'\n'}) {
			if len(bytes.TrimSpace(line)) == 0 {
				continue
			}
			if !json.Valid(line) {
				return nil, errors.New("invalid Bench stream event")
			}
			var event struct {
				Type, Code, Message string
				Error               any
			}
			_ = json.Unmarshal(line, &event)
			if event.Type == "error" {
				if event.Code == "" {
					event.Code = "stream_error"
				}
				if event.Message == "" {
					event.Message = fmt.Sprint(event.Error)
				}
				return nil, &PlatformError{Status: 200, Code: event.Code, Message: event.Message}
			}
			events = append(events, append(json.RawMessage(nil), line...))
		}
		return json.Marshal(events)
	}
	if !json.Valid(raw) {
		return nil, errors.New("invalid Bench JSON response")
	}
	return raw, nil
}
